### PR TITLE
Add static and dynamic port overrides for CNI ebpf

### LINKED ISSUE #9817: Add port `6443` to the default list of skipped ports for control plane components
It is possible for the control plane to get stuck when installing Linkerd, if the environment makes use of a CNI plugin that replaces `kube-proxy` with ebpf based solutions (e.g #8591).

Cilium (and other ebpf-based CNIs) can do socket-level load balancing, which has the downside of changing the target from the `ClusterIP` to the actual backend during the tcp `connect` syscall.

> sockets are assigned to service backends directly, e.g. at TCP connect time [[1]](https://docs.cilium.io/en/v1.9/gettingstarted/kubeproxy-free/#maglev-consistent-hashing-beta)

By default, the Kubernetes api server listens on `6443`, this can be changed by users when clusters are rolled out, but should be pretty consistent across managed distributions. The default is documented in the [official reference](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-apiserver/). 

```
--secure-port int     Default: 6443
  | The port on which to serve HTTPS with authentication and authorization. It cannot be switched off with 0.
```

**The implication for Linkerd's control plane**: containers (e.g identity, destination) will talk directly to _6443_, instead of 443, which is not skipped by default. This requires manual patching as the control plane is not injected automatically. We should include `6443` in the list of default skipped ports. This would require us to change the CNI plugin source code (which hardcodes 443 as a skipped port for control plane components), and the helm charts.

### PR DESCRIPTION
**Pull request by mateiidavid**
When CNI plugins run in ebpf mode, they may rewrite the packet destination when doing socket-level load balancing (i.e in the `connect()` call). In these cases, skipping `443` on the outbound side for control plane components becomes redundant; the packet is re-written to target the actual Kubernetes API Server backend (which typically listens on port `6443`, but may be overridden when the cluster is created).

This change adds port `6443` to the list of skipped ports for control plane components. On the linkerd-cni plugin side, the ports are non-configurable. Whenever a pod with the control plane component label is handled by the plugin, we look-up the `kubernetes` service in the default namespace and append the port values (of both ClusterIP and backend) to the list.

On the initContainer side, we make this value configurable in Helm and provide a sensible default (`443,6443`). Users may override this value if the ports do not correspond to what they have in their cluster. In the CLI, if no override is given, we look-up the service in the same way that we do for linkerd-cni; if failures are encountered we fallback to the default list of ports from the values file.

Closes #9817

Signed-off-by: Matei David <matei@buoyant.io>

---

### Manually testing the changes

* Instructions to spin-up a k3d cluster running calico in ebpf mode

```
k3d cluster create calico-ebpf \
              --k3s-arg "--disable-network-policy@server:*" \
              --k3s-arg "--flannel-backend=none@server:*"

docker exec k3d-calico-ebpf-server-0 sh -c \
       'mount bpffs /sys/fs/bpf -t bpf;\
        mount --make-shared /sys/fs/bpf;\
        mount --make-rshared /'

# Will install operator
# then Calico
# we do a patch to make it run in ebpf mode
kubectl create -f https://raw.githubusercontent.com/projectcalico/calico/v3.24.5/manifests/tigera-operator.yaml
kubectl create -f https://raw.githubusercontent.com/projectcalico/calico/v3.24.5/manifests/custom-resources.yaml
kubectl patch installation.operator.tigera.io default --type merge -p '{"spec":{"calicoNetwork":{"linuxDataplane":"BPF", "hostPorts":null}}}'
```

**Linkerd CNI test**:

* Install `linkerd-cni` and linkerd with `--set 'enableCNI=true'`

```
; kgp -n linkerd
NAME                                      READY   STATUS    RESTARTS   AGE
linkerd-identity-548df67b4-gbhtf          2/2     Running   0          2m15s
linkerd-proxy-injector-6ff949ff54-6l4zf   2/2     Running   0          2m15s
linkerd-destination-77d45b86c8-kv4hw      4/4     Running   0          2m15s

# To verify iptables, had to exec on the k3d node, and do `ps`
# grab the PID of a `controller` process and then use `ip netns identify <pid>`
# followed by `ip netns exec <name> iptables-nft-save
11794 2103     /controller destination -addr=:8086 -controller-namespace=linkerd -enable-h2-upgrade=true -log-level=info -log-format=plain -
11839 2103     /controller sp-validator -log-level=info -log-format=plain -enable-pprof=false
12072 2103     /bin/linkerd-policy-controller --admin-addr=0.0.0.0:9990 --control-plane-namespace=linkerd --grpc-addr=0.0.0.0:8090 --server-
13094 0        ps
/var/log # ip netns identify 11794
cni-4ad4124b-0d00-aa1b-8d51-9a1acbc8f740
/var/log # ip netns exec cni-4ad4124b-0d00-aa1b-8d51-9a1acbc8f740 iptables-nft-save
# Generated by iptables-nft-save v1.8.6 on Wed Nov 16 15:57:43 2022
*nat
-A PROXY_INIT_OUTPUT -p tcp -m multiport --dports 443,6443 -m comment --comment "proxy-init/ignore-port-443,6443/1668614201" -j RETURN
# ^ confirms both ports were added to skip on outbound side
```

**Linkerd init container CLI**

* With the CLI, we can override values, and no default or look-up result will be applied:

```
:; bin/linkerd install --set 'proxyInit.kubeAPIServerPorts=333' | rg '333'
      kubeAPIServerPorts: 333
        - "333"
        - "333"
```

* I modified the `values.yaml` file to see if the look-up will override the default value when it doesn't correspond to what we have on the service side.

```
# Manually modified the value to be `443` instead of `443,6443`
:; rg 'kubeAPIServerPorts' charts/linkerd-control-plane/values.yaml
197:  kubeAPIServerPorts: "443"

:; bin/linkerd install | rg '443,6443'
      kubeAPIServerPorts: 443,6443
        - "443,6443"
        - "443,6443"

# Ports modified correctly based on look-up result ^
```

* Installed linkerd `2.12.2` to check if we have an error, and then replaced with the branch build to see if it is fixed, all steps pasted below.

```
:; kgp -n linkerd
NAME                                     READY   STATUS            RESTARTS   AGE
linkerd-destination-66fd44dc57-q2hfb     0/4     PodInitializing   0          12s
linkerd-proxy-injector-9f8764675-rnp8s   0/2     PodInitializing   0          12s
linkerd-identity-758d6bd889-h7xlq        0/2     Error             0          12s

:; k logs -n linkerd linkerd-identity-758d6bd889-h7xlq | tail -n 1
Defaulted container "identity" out of: identity, linkerd-proxy, linkerd-init (init)
time="2022-11-16T16:37:00Z" level=fatal msg="Failed to initialize identity service: Post \"https://10.43.0.1:443/apis/authorization.k8s.io/v1/selfsubjectaccessreviews\": net/http: TLS handshake timeout"

:; k logs -n linkerd linkerd-identity-758d6bd889-h7xlq linkerd-proxy | tail -n 1
[    17.815090s]  WARN ThreadId(02) identity:controller{addr=localhost:8080}:endpoint{addr=127.0.0.1:8080}: linkerd_reconnect: Failed to connect error=Connection refused (os error 111)

:; linkerd version
Client version: stable-2.12.2
Server version: stable-2.12.2

# Uninstalled before proceeding with re-install

:; bin/linkerd install | k apply -f - 

:; kgp -n linkerd
NAME                                     READY   STATUS    RESTARTS   AGE
linkerd-identity-9b668984c-4htf7         2/2     Running   0          6m
linkerd-destination-5f846595f7-ntmgn     4/4     Running   0          6m
linkerd-proxy-injector-b7777b9b8-dr7ld   2/2     Running   0          6m

# yaml snippet from identity
 initContainers:
  - args:
    - --incoming-proxy-port
    - "4143"
    - --outgoing-proxy-port
    - "4140"
    - --proxy-uid
    - "2102"
    - --inbound-ports-to-ignore
    - 4190,4191,4567,4568
    - --outbound-ports-to-ignore
    - 443,6443
```

**Not tested manually**:
* Upgrades
* Helm installs.

### CODE REVIEW

File: charts/linkerd-control-plane/values.yaml [hunk 1]
```diff
@@ -192,6 +192,9 @@ proxyInit:
   # -- Default set of outbound ports to skip via iptables
   # - Galera (4567,4568)
   ignoreOutboundPorts: "4567,4568"
+  # -- Default set of ports to skip via iptables for control plane
+  # -- components so they can communicate with the Kubernetes API Server
```
Comment (kleimkuhler) [inline charts/linkerd-control-plane/values.yaml:196]: It looks like this makes the comment separate.

```suggestion
  # components so they can communicate with the Kubernetes API Server
```

Comment (mateiidavid) [reply charts/linkerd-control-plane/values.yaml:196]: Nice, good spot!

File: cli/cmd/install.go [hunk 2]
```diff
@@ -227,6 +230,14 @@ func installControlPlane(ctx context.Context, k8sAPI *k8s.KubernetesAPI, w io.Wr
 				os.Exit(1)
 			}
 		}
+
+		// Check 'kubernetes' service in default namespace to see what ports the API
+		// Server listens on. If the ports are different from the default ('443,6443')
+		// then replace with ports from the service spec.
+		apiSrvPorts := getApiServerPorts(ctx, k8sAPI)
+		if apiSrvPorts != "" && apiSrvPorts != values.ProxyInit.KubeAPIServerPorts {
```
[the remark this answers is not in the document]
Comment (mateiidavid) [reply cli/cmd/install.go:238]: Yeah, you're right, it's redundant. Good catch.

File: cni-plugin/main.go [hunk 3]
```diff
@@ -249,9 +250,18 @@ func cmdAdd(args *skel.CmdArgs) error {
...
-				// Skip 443 outbound port if its a control plane component
-				logEntry.Debug("linkerd-cni: adding 443 to OutboundPortsToIgnore as its a control plane component")
-				options.OutboundPortsToIgnore = append(options.OutboundPortsToIgnore, "443")
+				// Skip k8s api server ports on the outbound side if pod is a
+				// control plane component
+				skippedPorts, err := getApiServerPorts(ctx, client)
+				if err != nil {
+					// If we cannot retrieve the 'kubernetes' service's ports (for
+					// whatever reason), skip default ports: 443, 6443
+					logEntry.Debugf("linkerd-cni: could not retrieve ports from 'kubernetes' service: %v", err)
```
Comment (adleong) [inline cni-plugin/main.go:259]: probably worth logging this at Error level instead of Debug.

**Review by adleong — APPROVED** [review]
One suggestion about log level, otherwise looks great!

File: cni-plugin/main.go (hunk 3 above)
Comment (mateiidavid) [reply cni-plugin/main.go]: Nice, done!

**Review by stevej — APPROVED** [review]

**Review by kleimkuhler — APPROVED** [review]

**Comment by mateiidavid** [comment]
Thanks all for the reviews

**Pull request merged**